src/deno_dir.rs

```diff
@@ -136,38 +137,81 @@ impl DenoDir {
-        Some(ref parent) => fs::create_dir_all(parent),
-        None => Ok(()),
-      }?;
-      deno_fs::write_file(&p, source.as_bytes(), 0o666)?;
-      deno_fs::write_file(&mt, content_type.as_bytes(), 0o666)?;
-      (source, map_content_type(&p, Some(&content_type)))
-    } else {
-      let source = fs::read_to_string(&p)?;
-      // .mime file might not exists with bundled deps
+  ) -> DenoResult<Option<CodeFetchOutput>> {
+    let extensions = ["", ".ts", ".js"];
+    for ext in extensions.iter() {
+      let filename = format!("{}{}", filename, ext);
+      let module_name = format!("{}{}", module_name, ext);
+      let p = Path::new(&filename);
+      // We write a special ".mime" file into the `.deno/deps` directory along side the
+      // cached file, containing just the media type.
+      let mut media_type_filename = filename.to_string();
+      media_type_filename.push_str(".mime");
+      let mt = Path::new(&media_type_filename);
+      eprint!("Downloading {}...", &module_name); // no newline
+      let maybe_source = http_util::fetch_sync_string(&module_name);
+      if let Ok((source, content_type)) = maybe_source {
+        eprintln!(""); // next line
+        match p.parent() {
+          Some(ref parent) => fs::create_dir_all(parent),
+          None => Ok(()),
+        }?;
+        deno_fs::write_file(&p, source.as_bytes(), 0o666)?;
+        deno_fs::write_file(&mt, content_type.as_bytes(), 0o666)?;
+        return Ok(Some(CodeFetchOutput {
+          module_name: module_name.clone(),
+          filename: filename.clone(),
+          media_type: map_content_type(&p, Some(&content_type)),
+          source_code: source,
+          maybe_output_code: None,
+          maybe_source_map: None,
+        }));
+      } else {
+        eprintln!(" NOT FOUND");
+      }
+    }
+    Ok(None)
+  }
+
+  /// Fetch local or cached source code.
+  fn fetch_local_source(
+    self: &Self,
+    module_name: &str,
+    filename: &str,
+  ) -> DenoResult<Option<CodeFetchOutput>> {
+    let extensions = ["", ".ts", ".js"];
+    for ext in extensions.iter() {
+      let filename = format!("{}{}", filename, ext);
+      let module_name = format!("{}{}", module_name, ext);
+      let p = Path::new(&filename);
+      if !p.exists() {
+        continue;
+      }
+      let mut media_type_filename = filename.to_string();
+      media_type_filename.push_str(".mime");
+      let mt = Path::new(&media_type_filename);
+      let source_code = fs::read_to_string(&p)?;
+      // .mime file might not exists
+      // this is okay for local source: maybe_content_type_str will be None
       let maybe_content_type_string = fs::read_to_string(&mt).ok();
       // Option<String> -> Option<&str>
       let maybe_content_type_str =
         maybe_content_type_string.as_ref().map(String::as_str);
-      (source, map_content_type(&p, maybe_content_type_str))
-    };
-    Ok(src)
+      return Ok(Some(CodeFetchOutput {
+        module_name: module_name.clone(),
```

Comment:
The `clone()` seems unnecessary?

Reply:
Hmm `module_name.clone()` is unnecessary, but `filename.clone()` is still needed

Reply:
Also I'll make a final push after you complete review... (travis is jammed)

Reply:
why? You have created a local `filename : String` which shadows `filename: &str` in the argument.

Reply:
`let mt = Path::new(&media_type_filename);`

Reply:
`p` is inactive after its last usage. It should compile when NLL is enabled. Before that, we can only bypass this restriction (annoying):
```rust
let filename = ...
{
  let p = Path::new(&filename);
  ...
}
// move filename
```

Reply:
Add a TODO ?

Reply:
Err, `map_content_type(&p, Some(&content_type))` should also move to  another place... Yeah, it depends on you whether to choose this style.

Reply:
@F001 I'll drop it and add a TODO